[ArtistFilter] Working unstyled component w/ inf scroll

mzikherman opened the pull request:
![bp](https://user-images.githubusercontent.com/1457859/41001261-752d458c-68de-11e8-84f7-8b2f9f384968.gif)

This has a quick pass at a somewhat functional filter component for the artist page. It uses `unstated` for state management (so no `setState` appears in our components).

There's a fragment container for the filter sidebar and artworks, and the artworks 'browser' is a pagination container. The prev/next windowed pagination appears to be quite tricky...I haven't found any good libs for that yet, so didn't really make too much of a stab at that yet (besides a hacky 'next' button). I'll make some comments inline.

Depends on some local Metaphysics changes as well, will open a PR there.

mzikherman commented on src/Components/ArtistFilter/Artworks.tsx:
```diff
@@ -0,0 +1,147 @@
+          return (
+            <div>
+              {this.renderPagination(filters)}
+              <ArtworkGrid
```
Just uses existing `ArtworkGrid` in order to render the artworks.

mzikherman commented on src/Components/ArtistFilter/Artworks.tsx:
```diff
@@ -0,0 +1,147 @@
+    const endCursor = this.props.filtered_artworks.artworks.pageInfo.endCursor
+    if (hasMore) {
+      // TODO: Should refetchConnection keep appending records?
+      this.props.relay.refetchConnection(
```
It seems like this should keep re-initializing the connection, but in testing this seems to just keep appending records (so analogous to `loadMore`), leading to a hack to only display the current page of artworks.

We'll probably need to come back to this (good call at identifying that prev/next as well as arbitrary page number jumping is going to be difficult).

damassi replied:
@mzikherman - introducing the `paginationContainer` would fix this, yah?

mzikherman commented on src/Components/ArtistFilter/Filter.tsx:
```diff
@@ -0,0 +1,164 @@
+    return state[filter]
+  }
+
+  renderFilters(filters) {
```
This renders all our filters, and displays the currently selected filters (if there are any).

mzikherman commented on src/Components/ArtworkGrid.tsx:
```diff
@@ -57,8 +59,11 @@ class ArtworkGridContainer extends React.Component<Props, State> {
-    const artworks = this.props.artworks ? this.props.artworks.edges : []
+    let artworks = this.props.artworks ? this.props.artworks.edges : []
 
+    if (this.props.page && this.props.pageSize) {
```
This is the hack to only display the current page of works in the grid.

alloy commented:
@mzikherman I think that for windowed pagination you should use a refetch container (which changes the page parameter) instead of a pagination container.

alloy commented on src/Components/ArtistFilter/Artworks.tsx:
```diff
@@ -0,0 +1,147 @@
+import { FilterState } from "./state"
+import ArtworkGrid from "../ArtworkGrid"
+
+interface Props extends RelayProps {
```
You don’t need to write and `RelayProps` anymore. Instead import the TS typings emitted by our relay-compiler TS plugin, in this case you should import from [this file](https://github.com/artsy/reaction/blob/d31a2db8ff9bc55e1eaba8d88ad8ccbf556f27f5/src/__generated__/Artworks_filtered_artworks.graphql.ts):

```ts
import { Artworks_filtered_artworks } from "../../__generated__/Artworks_filtered_artworks.graphql"

interface Props {
  filtered_artworks: Artworks_filtered_artworks
}
```

alloy commented on src/Components/ArtistFilter/Artworks.tsx:
```diff
@@ -0,0 +1,147 @@
+import ArtworkGrid from "../ArtworkGrid"
+
+interface Props extends RelayProps {
+  relay?: RelayPaginationProp
```
This shouldn’t need to be optional anymore, the external props interface of the relay container won’t include/require this prop anymore with the current DefinitelyTyped typings.

damassi commented on src/Components/ArtistFilter/Artworks.tsx:
```diff
@@ -0,0 +1,147 @@
+
+const PageSize = 10
+
+class Artworks extends React.Component<Props, null> {
```
Can omit `null` if not using state

damassi commented on src/Components/ArtistFilter/Artworks.tsx:
```diff
@@ -0,0 +1,147 @@
+  artistID: string
+}
+
+const PageSize = 10
```
This is likely taken from elsewhere in codebase but its more idiomatic to see constants like this expressed as 

```javascript
const PAGE_SIZE = 10
```

mzikherman commented:
Alright, I think this is mergeable, as the changes are all in the new component (I removed the hacky `ArtworkGrid` pagination).

I left the artworks content as `createPaginationContainer` and inf scroll (since that basically works out of the box), and added `page` to state. Next up I can work on a more proper pagination control (basically `refetchContainer` with adding support for `page` `size` and `totalPages` in the filter artworks schema). This should enable functional paging.
